### PR TITLE
Fix new DBML parser inconsistencies

### PR DESCRIPTION
**Pull request by NQPhuc**
### Resolve the following inconsistencies between old and new dbml parser behaviors:
1. Fix a bug causing multiple single quote to not report error.
```
Table user_role_in_diagram {
  role int [note: 'Role = sum(user's available permissions bit value)'] // should throw error on 's
}
```
2. Make keyworks check for default settings (true, false and null) case-insensitive.
3. Allow default value with double quote (https://github.com/holistics/dbml/issues/481)

4. Allow inline keywords in columns.
```
Table t {
  id int pk
}
```
5. Allow multiple column names on the same line in indexes definition.
```
Table citites {
  id integer pk
  name string

  indexes {
    id name 
  }
}
```
6. Allow array type in column.
```
Table users {
  id int
  order_ids int[]
}
```

* Improve error messages on validation errors
* Fix some errors reported with invalid error location (NaN on line and column)

## Issue
(issue link here)

## Lasting Changes (Technical)
* Ignore null attribute and setting bug cause by syntax error involve more than 2 single quotes
* Stop binding default indentifier to Enum
* Keywords check in binder and intepreter is now case-insensivity
* Update tests
* Refactor validator to not use the Template Method pattern
* Refactor Interpreter to make it modular

## Checklist

Please check directly on the box once each of these are done

- [ ] Documentation (if necessary)
- [x] Tests (integration test/unit test)
- [x] Integration Tests Passed
- [ ] Code Review

### CODE REVIEW

**Comment by HuyDNA** [comment]
> 3. Unsupported: Array column types — due to the ambiguity in the grammar that this would introduce.
> 
> ```
> Table country {
>   id integer
>   cities string[] // not work
>   // equivalent: cities "string[]"
> }
> ```
This has been supported. However, there's a pitfall:

```
Table country[note: 'abc'] {
  ...
}
```
`country[note: 'abc']` would be treated as an array node, instead of separately as a name and a list, due to some ambiguity.
This can be avoided by adding a space between them.
It can indeed be solved completely but it would complicate the parser significantly. In my opinion, this error is unlikely to happen and we can yield more informative error messages for this pitfall so the user can be more aware? @NQPhuc

**Comment by NQPhuc** [comment]
@HuyDNA I'm aware that since both table definition and column definition used the same rule (element-declaration), supporting array type for column would affect table definition as well.

I think for now we can accept the new inconsistency cause to table setting. **However, please help show a informative message to let the user know they need to leave empty space between table name and setting [].**

In the long run, we should consider breaking element-declaration into different declaration type depending, because as of right now, this is not very flexible.

**Comment by HuyDNA** [comment]
@NQPhuc All consistencies have been resolved. Plus the following benefits:
* Error messages are more informative
* Better error markings: Example, all duplicate settings are marked, instead of only the last; All circular refs are marked, instead of only the last, etc. -> Better debugging
* The Interpreter is refactored into modules instead of being monolithic, adding other elements is easier.
* The Validator is more flexible, allowing catching more precise errors.

File: packages/dbml-parse/src/lib/analyzer/validator/utils.ts [hunk 1]
```diff
@@ -25,47 +27,26 @@ import { createSchemaSymbolIndex } from '../symbol/symbolIndex';
...
-    case ElementKind.PROJECT:
+    case 'project':
       return ProjectValidator;
-    case ElementKind.REF:
+    case 'ref':
       return RefValidator;
-    case ElementKind.NOTE:
+    case 'note':
       return NoteValidator;
-    case ElementKind.INDEXES:
```
Comment (NQPhuc) [inline packages/dbml-parse/src/lib/analyzer/validator/utils.ts:68]: @HuyDNA Please help re-add this constant nha. We shouldn't fixed the element type somewhere deep in the validator code.

File: packages/dbml-parse/src/lib/analyzer/validator/elementValidators/custom.ts [hunk 2]
```diff
@@ -1,74 +1,75 @@
...
+  }
+
+  validate(): CompileError[] {
+    return [...this.validateContext(), ...this.validateName(this.declarationNode.name), ...this.validateAlias(this.declarationNode.alias), ...this.validateSettingList(this.declarationNode.attributeList), ...this.validateBody(this.declarationNode.body)];
+  }
 
-export default class CustomValidator extends ElementValidator {
-  protected elementKind: ElementKind = ElementKind.CUSTOM;
+  private validateContext(): CompileError[] {
+    if (this.declarationNode.parent instanceof ProgramNode || this.declarationNode.parent?.type?.value.toLowerCase() !== 'project') {
```
Comment (NQPhuc) [inline packages/dbml-parse/src/lib/analyzer/validator/elementValidators/custom.ts:25]: Please use ElementKind here

File: packages/dbml-parse/src/lib/analyzer/validator/elementValidators/indexes.ts [hunk 3]
```diff
@@ -1,88 +1,172 @@
...
+    this.publicSymbolTable = publicSymbolTable;
+    this.symbolFactory = symbolFactory;
+  }
+
+  validate(): CompileError[] {
+    return [...this.validateContext(), ...this.validateName(this.declarationNode.name), ...this.validateAlias(this.declarationNode.alias), ...this.validateSettingList(this.declarationNode.attributeList), ...this.validateBody(this.declarationNode.body)];
+  }
+
+  private validateContext(): CompileError[] {
+    if (this.declarationNode.parent instanceof ProgramNode || this.declarationNode.parent?.type?.value.toLowerCase() !== 'table') {
```
Comment (NQPhuc) [inline packages/dbml-parse/src/lib/analyzer/validator/elementValidators/indexes.ts:38]: Also here

File: packages/dbml-parse/src/lib/interpreter/elementInterpreter/project.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+          const errors = (new TableGroupInterpreter(sub, this.env)).interpret();
+          this.project.tableGroups!.push(this.env.tableGroups.get(sub)!);
+          return errors;
+        }
+        case 'enum': {
+          const errors = (new EnumInterpreter(sub, this.env)).interpret();
+          this.project.enums!.push(this.env.enums.get(sub)!);
+          return errors;
+        }
+        case 'note': {
```
Comment (NQPhuc) [inline packages/dbml-parse/src/lib/interpreter/elementInterpreter/project.ts:64]: Should use elementKind here

**Review by huyphung1602 — APPROVED** [review]

**Pull request merged**